### PR TITLE
Parsing and emitting nominal types

### PR DESCRIPTION
**Pull request by tlively**
Adds a `--nominal` option to switch the type machinery from equirecursive to
nominal. Implements binary and text parsing and emitting of nominal types using
new type constructor opcodes and an `(extends $super)` text syntax extension.
When not in nominal mode, these extensions will still be parsed but will not
have any effect and will not be used when emitting.

### CODE REVIEW

File: test/lit/nominal-chain.wast [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+;; types.
+
+(module
+
+  ;; CHECK:      (type $root (struct ))
+  ;; CHECK-NEXT: (type $none_=>_ref?|$root| (func (result (ref null $root))))
+  ;; CHECK-NEXT: (type $leaf (struct (field i32) (field i64) (field f32) (field f64)) (extends $twig))
+  ;; CHECK-NEXT: (type $twig (struct (field i32) (field i64) (field f32)) (extends $branch))
+  ;; CHECK-NEXT: (type $branch (struct (field i32) (field i64)) (extends $trunk))
+  ;; CHECK-NEXT: (type $trunk (struct (field i32)) (extends $root))
```
Comment (kripken) [inline test/lit/nominal-chain.wast:15]: did you add these manually? will the auto-update script not trample them later? (we don't test the script on CI)

File: src/wasm-type.h [hunk 2]
```diff
@@ -363,6 +363,8 @@ class HeapType {
   const Struct& getStruct() const;
   Array getArray() const;
 
+  bool getSuperType(HeapType& out) const;
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/wasm-type.h:366]: We don't have a `HeapType` equivalent of `Type::none`, unfortunately. If we had C++17, I would use `std::optional<HeapType>`.

File: src/wasm/wasm-validator.cpp [hunk 3]
```diff
@@ -2508,6 +2508,9 @@ void FunctionValidator::visitFunction(Function* curr) {
   }
   // if function has no result, it is ignored
   // if body is unreachable, it might be e.g. a return
+  if (!Type::isSubType(curr->body->type, curr->sig.results)) {
+    std::cerr << curr->body->type << "\n" << curr->sig.results << "\n";
+  }
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/wasm/wasm-validator.cpp:2513]: Oops, this was for debugging. Thanks!

File: test/lit/nominal-chain.wast (hunk 1 above)
Comment (tlively) [reply test/lit/nominal-chain.wast:15]: The checks in this file are not generated by the auto-update script, so it shouldn't be an issue.

**Review by kripken — APPROVED** [review]

**Comment by kripken** [comment]
This caused some fuzzer breakage, see https://github.com/WebAssembly/binaryen/pull/3935#issuecomment-868816085 (mentioned there as I erroneously commented there initially.)

**Comment by tlively** [comment]
Thanks for diagnosing that. I'll investigate 👍

**Pull request merged**